R/derive_locf_records.R

```diff
@@ -0,0 +1,184 @@
+#'   "CDISC01", "01-701-1015", "SYSBP", "Systolic Blood Pressure (mmHg)", 121,    6,       "WEEK 6",
+#'   "CDISC01", "01-701-1028", "DIABP", "Diastolic Blood Pressure (mmHg)", 79,    0,       "BASELINE",
+#'   "CDISC01", "01-701-1028", "DIABP", "Diastolic Blood Pressure (mmHg)", 80,    2,       "WEEK 2",
+#'   "CDISC01", "01-701-1028", "DIABP", "Diastolic Blood Pressure (mmHg)", NA,    4,       "WEEK 4",
+#'   "CDISC01", "01-701-1028", "DIABP", "Diastolic Blood Pressure (mmHg)", NA,    6,       "WEEK 6",
+#'   "CDISC01", "01-701-1028", "SYSBP", "Systolic Blood Pressure (mmHg)", 130,    0,       "BASELINE",
+#'   "CDISC01", "01-701-1028", "SYSBP", "Systolic Blood Pressure (mmHg)", 132,    2,       "WEEK 2"
+#' )
+#'
+#'
+#' # A dataset with all the combinations of PARAMCD, PARAM, AVISIT, AVISITN, ... which are expected.
+#' advs_expected_obsv <- tibble::tribble(
+#'  ~PARAMCD, ~PARAM, ~AVISITN, ~AVISIT,
+#'   "DIABP", "Diastolic Blood Pressure (mmHg)", 0, "BASELINE",
+#'   "DIABP", "Diastolic Blood Pressure (mmHg)", 2, "WEEK 2",
+#'   "DIABP", "Diastolic Blood Pressure (mmHg)", 4, "WEEK 4",
+#'   "DIABP", "Diastolic Blood Pressure (mmHg)", 6, "WEEK 6",
+#'   "SYSBP", "Systolic Blood Pressure (mmHg)",  0, "BASELINE",
+#'   "SYSBP", "Systolic Blood Pressure (mmHg)",  2, "WEEK 2",
+#'   "SYSBP", "Systolic Blood Pressure (mmHg)",  4, "WEEK 4",
+#'   "SYSBP", "Systolic Blood Pressure (mmHg)",  6, "WEEK 6"
+#' )
+#'
+#' # Add a new record for each 'by_vars' when the input dataset does not contain
+#' # observations for missed visits/timepoints or when AVAL is missing.
+#' # Set DTYPE = 'LOCF' for the newly added records.
+#' # Carry forward the AVAL of the previous observation after sorting the dataset
+#' # by 'order' parameter.
+#' Final_advs <- derive_locf_records(data=advs,
+#'                                   dataset_expected_obs=advs_expected_obsv,
+#'                                   by_vars=vars(STUDYID, USUBJID, PARAM, PARAMCD),
+#'                                   order=vars(STUDYID, USUBJID, PARAM, PARAMCD, AVISITN, AVISIT))
+
+
+
+derive_locf_records <- function(dataset,
+                                dataset_expected_obs,
+                                by_vars,
+                                order) {
+  library(tidyr)
+
+  #### Input Checking ####
+
+  # Check by_vars and order variables in input dataset
+  assert_data_frame(dataset, required_vars = by_vars)
+  assert_data_frame(dataset, required_vars = order)
+
+  # Check if input parameters is a valid list of variables
+  assert_vars(by_vars, optional = TRUE)
+  assert_order_vars(order)
+
+
+
+  #### Prepping 'dataset_expected_obs' ####
+
+  # Add temp_dtype='LOCF' for the 'expected observation dataset'
+  dataset_expected_obs <- dataset_expected_obs %>%
+    mutate(temp_dtype="LOCF")
+
+
+  # Get the STUDYID and USUBJID from input dataset to set expected
+  # observations for each subject
+  subjids <- dataset %>%
+    select(STUDYID, USUBJID) %>%
```

Comment:
`by_vars` would not work in this case, as I'm trying to get the observations in `dataset_expected_obs` for each STUDYID, USUBJID. So, I was wondering if this would work - 
The user directly passes a dataset in `dataset_expected_obs` that has the expected observations for each STUDYID, USUBJID. Then the part of the function performing this operation would not be required.

Reply:
I see. I think it should be
```
  select(setdiff(by_vars, chr2vars(colnames(dataset_expected_obs))))
```
If you always use `STUDYID`, `USUBJID` the following does not work.
```
advs_expected_obsv <- tibble::tribble(
   ~AVISITN, ~AVISIT,
   0, "BASELINE",
   2, "WEEK 2",
   4, "WEEK 4",
   6, "WEEK 6"
 )
derive_locf_records(
   data = advs,
   dataset_expected_obs = advs_expected_obsv,
   by_vars = vars(STUDYID, USUBJID, PARAMCD),
   order = vars(STUDYID, USUBJID, PARAMCD, AVISITN, AVISIT)
 )
```

Reply:
Alright. I've removed `STUDYID` and `USUBJID`. But this was throwing as error:
`select(setdiff(by_vars, chr2vars(colnames(dataset_expected_obs))))`

So I've added this, which works fine:
`select(setdiff(str_remove(as.character(by_vars), '~'), colnames(dataset_expected_obs)))`